### PR TITLE
perf(core): add private hooks around user code executed by the runtime

### PR DESCRIPTION
**Pull request by mgechev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?

Introduces a `profiler` function, which the runtime invokes around user
code, including before and after:
- Running the template function of a component
- Executing a lifecycle hook
- Evaluating an output handler

The `profiler` function invokes a callback set with the global `ng.ɵsetProfiler`.

This implementation is cheap and available in production. It's cheap because the `profiler` function is simple, which allows the JiT compiler to inline it in the callsites. It also doesn't add up much to the production bundle.

To listen for profiler events:

```ts
ng.ɵsetProfiler((event, ...args) => {
  // monitor user code execution
});
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/render3/profiler.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * @param event ProfilerEvent corresponding to the execution context
+ * @param instanceOrLView component instance or corresponding lView. The value depends on the
+ *  execution context
+ * @param hookOrListener lifecycle hook function or output listener. The value depends on the
+ *  execution context
+ * @returns
+ */
+export const profiler: Profiler = function(event: any, instanceOrLView: any, hookOrListener?: any) {
+  if (!globalNg || !globalNg.ɵprofiler) {
```
Comment (gkalpak) [inline packages/core/src/render3/profiler.ts:51]: Do we want to support updating `globalNg.ɵprofiler` at runtime?
(Otherwise, wouldn't it be more performant to set `const profiler` to a no-op function if `globalNg.ɵprofiler` is not defined?)

Comment (petebacondarwin) [reply packages/core/src/render3/profiler.ts:51]: I assume that if `globalNg.ɵprofiler` is a truthy constant then terser (or other minifier) would be able to see that this always returns early and drop the following line, making it a noop; and then inline the function, which will effectively remove it altogether.

Comment (gkalpak) [reply packages/core/src/render3/profiler.ts:51]: You mean "fasly constant"?
But I don't think terser can (or should) do that, because `globalNg.ɵprofiler` is not a built-time constant (it is a runtime value). No?

Comment (gkalpak) [reply packages/core/src/render3/profiler.ts:51]: My understanding is that it is not, but I could be wrong 😅

Comment (mgechev) [reply packages/core/src/render3/profiler.ts:51]: >My understanding is that it is not, but I could be wrong 😅

I have the same understanding. I don't think terser would be able to detect that's dead code. I'm setting profiler to `null` just to ensure consistent shape of `ng`, which shouldn't have significant impact so if you prefer I could just drop it.

Another approach could be to wrap everything with `ngDevMode` to ensure the code does not get into the production bundle, but at certain point we'd want to make the functionality available for prod apps as well. Based on discussions with @mhevery, seems like there should be no runtime overhead because the JiT compiler of the VM will inline `profiler` into the call sites. The impact on the bundle size is also minimal.

>Do we want to support updating `globalNg.ɵprofiler` at runtime?

The instrumenting code will set `globalNg.ɵprofiler` dynamically and we can't guarantee it will execute before the framework code.

Comment (JoostK) [reply packages/core/src/render3/profiler.ts:51]: What if we'd have `globalNg.setProfiler(profiler: Profiler)` instead, which would bind into a top-level constant that is a noop function by default. That avoids the branch instructions for `!globalNg || !globalNg.ɵprofiler`.

File: packages/core/src/render3/instructions/shared.ts [hunk 2]
```diff
@@ -501,16 +502,23 @@ export function renderComponentOrTemplate<T>(
...
   try {
     setSelectedIndex(-1);
-    if (rf & RenderFlags.Update && lView.length > HEADER_OFFSET) {
+    if (update && lView.length > HEADER_OFFSET) {
       // When we're updating, inherently select 0 so we don't
       // have to generate that instruction for most update blocks.
       selectIndexInternal(tView, lView, HEADER_OFFSET, isInCheckNoChangesMode());
     }
+
+    const preHookType = update ? ProfilerEvent.ChangeDetectionStart : ProfilerEvent.CreationStart;
```
Comment (JoostK) [inline packages/core/src/render3/instructions/shared.ts:514]: It would be interesting to derive the `ProfilerEvent` from `update` to avoid the branch. `update` has a value of 0 or 2, so if `ProfilerEvent.ChangeDetectionStart` is 2 higher than `ProfilerEvent.CreationStart` you could use `const preHookType = ProfilerEvent.CreationStart + update` (and some casting to make that compile)

File: packages/core/src/render3/instructions/listener.ts [hunk 3]
```diff
@@ -229,11 +230,16 @@ function listenerInternal(
 function executeListenerWithErrorHandling(
     lView: LView, listenerFn: (e?: any) => any, e: any): boolean {
   try {
+    // Not passing directly the context because the compiler
+    // might not be able to efficiently inline the profiler.
+    profiler(ProfilerEvent.OutputStart, lView, listenerFn);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/listener.ts:235]: Given that we may want to expose this as a public API later, I think we should avoid passing internal data structures to the profiler hook and instead pass component instance, DOM element and other objects that developers interact with.

File: packages/core/src/render3/instructions/shared.ts [hunk 4]
```diff
@@ -501,16 +502,23 @@ export function renderComponentOrTemplate<T>(
 function executeTemplate<T>(
     tView: TView, lView: LView, templateFn: ComponentTemplate<T>, rf: RenderFlags, context: T) {
   const prevSelectedIndex = getSelectedIndex();
+  const update = rf & RenderFlags.Update;
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions/shared.ts:505]: May be rename this to indicate what the const represents?
```suggestion
  const isUpdatePhase = rf & RenderFlags.Update;
```

File: packages/core/test/render3/profiler_spec.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+import {AttributeMarker, getHostElement, getRenderedText, LifecycleHooksFeature, ɵɵadvance, ɵɵdefineComponent, ɵɵgetCurrentView, ɵɵloadQuery, ɵɵproperty, ɵɵqueryRefresh, ɵɵtextInterpolate1, ɵɵtextInterpolate2, ɵɵviewQuery} from '../../src/render3/index';
+import {detectChanges, markDirty, tick, ɵɵelement, ɵɵelementEnd, ɵɵelementStart, ɵɵlistener, ɵɵtemplate, ɵɵtext, ɵɵtextInterpolate} from '../../src/render3/instructions/all';
+import {RenderFlags} from '../../src/render3/interfaces/definition';
+import {Renderer3, RendererFactory3} from '../../src/render3/interfaces/renderer';
+import {FLAGS, LViewFlags} from '../../src/render3/interfaces/view';
+
+import {NgIf} from './common_with_def';
+import {containerEl, createComponent, renderComponent, requestAnimationFrame} from './render_util';
+
+describe('profiler', () => {
```
Comment (AndrewKushnir) [inline packages/core/test/render3/profiler_spec.ts:26]: We should consider re-creating these tests as acceptance ones and use TestBed API, see existing spec files in `packages/core/test/acceptance` folder.

File: packages/core/src/render3/instructions/shared.ts (hunk 2 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/shared.ts:514]: Provided that `ProfileEvent` is `const enum` i doubt this would result in noticeable perf difference.

File: packages/core/src/render3/profiler.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {global} from '../util/global';
+import {LView} from './interfaces/view';
+
+export const enum ProfilerEvent {
+  CreationStart,
```
Comment (mhevery) [inline packages/core/src/render3/profiler.ts:13]: These need solid jsdocs explaining when they get fired.

File: packages/core/src/render3/profiler.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/profiler.ts:51]: > What if we'd have `globalNg.setProfiler(profiler: Profiler)` instead, which would bind into a top-level constant that is a noop function by default. That avoids the branch instructions for `!globalNg || !globalNg.ɵprofiler`.

I like @JoostK  suggestion. We should go with that approach.

Notes:
- We don't want to have `noop` functions. Doing so would inline them into the callsite, which is great, but then a call to `setProfiler` would cause a deop as VM would realize that the inline was incorrect. This would result in the profiler would run slower.
- Having `if(fn) fn()` is better because the setting `fn` would change the `if` and it would not cause a deopt, it would only cause the branch prediction on `if` to change. Branch prediction is CPU concept (not VM concept)

File: packages/core/src/render3/instructions/listener.ts (hunk 3 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/listener.ts:235]: Thanks @mhevery, that makes sense.

I believe we could consider passing extra data from the `listenerInternal` function (which is executed at creation time only, so it's a bit less perf-sensitive) and just use it in profiler call. In `listenerInternal` function  we have access to the original event name (instead of deriving it from the listernerFn function name) and we can probably do `lView[CONTEXT]` read there too.

File: packages/core/src/render3/profiler.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+ */
+export const setProfiler = (profiler: Profiler) => {
+  profilerCallback = profiler;
+};
+
+/**
+ * Profiler function which wraps user code executed by the runtime.
+ *
+ * @param event ProfilerEvent corresponding to the execution context
+ * @param instanceOrLView component instance or corresponding lView. The value depends on the
```
Comment (AndrewKushnir) [inline packages/core/src/render3/profiler.ts:104]: I believe after refactoring in listener fn we don't have a situation when `LView` is passed?

File: packages/core/src/render3/util/global_utils.ts [hunk 8]
```diff
@@ -40,6 +41,7 @@ let _published = false;
 export function publishDefaultGlobalUtils() {
   if (!_published) {
     _published = true;
+    publishGlobalUtil('ɵsetProfiler', setProfiler);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/util/global_utils.ts:44]: I would also propose adding a warning here:

```suggestion
    /**
     * Warning: this function is *INTERNAL* and should not be relied upon in application's code.
     * The contract of the function might be changed in any release and/or the function can be removed completely.
     */
    publishGlobalUtil('ɵsetProfiler', setProfiler);
```

Would it be more convincing to call it something like `__setProfiler__`? :)

File: packages/core/test/acceptance/profiler_spec.ts [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+  class Profiler {
+    profile() {}
+  }
+
+  let profilerSpy: jasmine.Spy;
+
+  beforeEach(() => {
+    const profiler = new Profiler();
+    profilerSpy = spyOn(profiler, 'profile').and.callThrough();
+    setProfiler(profiler.profile);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:27]: We should also do a cleanup at the end of each test, so we don't leak profiler in other tests:

```
afterAll(() => setProfiler(null));
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+  describe('change detection hooks', () => {
+    it('should call the profiler for creation and change detection', () => {
+      @Component({selector: 'my-comp', template: '<button (click)="onClick()"></button>'})
+      class MyComponent {
+        onClick() {}
+      }
+
+      TestBed.configureTestingModule({declarations: [MyComponent]});
+      const myComp = TestBed.createComponent(MyComponent);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:38]: I think it's a bit more conventional in acceptance tests to have a `fixture` const:
```suggestion
      const fixture = TestBed.createComponent(MyComponent);
      const myComp = fixture. componentInstance;
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        const fixture = TestBed.createComponent(MyComponent);
+        myComp = fixture.componentInstance;
+        fixture.detectChanges();
+      }).toThrow();
+
+      expect(profilerSpy).toHaveBeenCalled();
+      const [creationStart, creationEnd] = profilerSpy.calls.all();
+
+      expect(creationStart.args[0]).toBe(ProfilerEvent.CreationStart);
+      expect(creationStart.args[1]).toBe(myComp!);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:81]: You could probably use some method like `toHaveBeenCalledWith` and `toHaveBeenCalledTimes` here and also use `argsFor`, for ex:

```suggestion
      profilerSpy.calls.argsFor(0).toBe([ProfilerEvent.CreationStart, myComp]);
      profilerSpy.calls.argsFor(1).toBe([ProfilerEvent.CreationEnd, myComp]);
```

File: packages/core/src/render3/instructions/shared.ts [hunk 12]
```diff
@@ -501,16 +502,25 @@ export function renderComponentOrTemplate<T>(
...
-    if (rf & RenderFlags.Update && lView.length > HEADER_OFFSET) {
+    if (isUpdatePhase && lView.length > HEADER_OFFSET) {
       // When we're updating, inherently select 0 so we don't
       // have to generate that instruction for most update blocks.
       selectIndexInternal(tView, lView, HEADER_OFFSET, isInCheckNoChangesMode());
     }
+
+    const preHookType =
+        isUpdatePhase ? ProfilerEvent.ChangeDetectionStart : ProfilerEvent.CreationStart;
+    profiler(preHookType, context);
```
[the remark this answers is not in the document]
Comment (mgechev) [reply packages/core/src/render3/instructions/shared.ts:516]: The children template functions will also be wrapped inside profiler calls so we'll get information about them.

Regarding view/content queries, even if we provide this information to people I'm not sure there's much they can do with it? Also the definition of the hook would be connected to when in the lifecycle of the app we invoke queries.

File: packages/core/src/render3/util/global_utils.ts (hunk 8 above)
Comment (mgechev) [reply packages/core/src/render3/util/global_utils.ts:44]: I kept it `ɵsetProfiler` for consistency with other APIs we use between tooling and framework.

File: packages/core/src/render3/profiler.ts [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+   * an event or an output.
+   */
+  OutputEnd,
+}
+
+/**
+ * Profiler function which the runtime will invoke before and after user code.
+ */
+export interface Profiler {
+  (event: ProfilerEvent.CreationStart|ProfilerEvent.CreationEnd|
```
Comment (mhevery) [inline packages/core/src/render3/profiler.ts:67]: why enumerate each of the events? Can't you just say `ProfilerEvent`?

File: packages/core/src/render3/profiler.ts [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+ * Profiler function which wraps user code executed by the runtime.
+ *
+ * @param event ProfilerEvent corresponding to the execution context
+ * @param instance component instance
+ * @param hookOrListener lifecycle hook function or output listener. The value depends on the
+ *  execution context
+ * @returns
+ */
+export const profiler: Profiler = function(
+    event: ProfilerEvent, instance: {}|null, hookOrListener?: (e?: any) => any) {
```
Comment (mhevery) [inline packages/core/src/render3/profiler.ts:97]: Why did we go away from `LView` and into context? Looking up `Lview` from context is expensive, whereas looking up `Context` from `Lview` is cheap.

**Review by mhevery — CHANGES_REQUESTED** [review]

File: packages/core/src/render3/profiler.ts (hunk 14 above)
Comment (mgechev) [reply packages/core/src/render3/profiler.ts:97]: I was addressing @AndrewKushnir's comment to not expose internals in the profiler. Having the `LViewDebug` being accessible from `LView` should probably mitigate this risk, given the `profiler` is still a private API?

Comment (AndrewKushnir) [reply packages/core/src/render3/profiler.ts:93]: My guess is that if we pass component instance to the `profiler` callback, there is no need to access the underlying `LView` anymore. @mgechev is that correct or you'd still need to get a reference to an `LView` in dev tools code?

Comment (mgechev) [reply packages/core/src/render3/profiler.ts:93]: Yes, no need to access the `LView` instance if we pass the component instance.

File: packages/core/src/render3/instructions/shared.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (AndrewKushnir) [reply packages/core/src/render3/instructions/shared.ts:516]: Yeah, invoking template function will probably take most of the time, however there is a couple things that might affect timing:
- the `refreshView` function also invokes host bindings
- query decorators might be applied to setters, which would have custom code too, for ex:

```

@ContentChild('someRef', {read: ElementRef})
set someField(value: ElementRef) {
  // ... some code to work with received `value` ...
}
```

File: packages/core/src/render3/util/global_utils.ts (hunk 8 above)
Comment (AndrewKushnir) [reply packages/core/src/render3/util/global_utils.ts:44]: Ok, makes sense.

What do you think of adding extra warnings into the code that the `ɵsetProfiler` should not be used in apps code (as mentioned in my comment above)?

File: packages/core/test/acceptance/profiler_spec.ts [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+      TestBed.configureTestingModule({declarations: [MyComponent]});
+      const fixture = TestBed.createComponent(MyComponent);
+      const myComp = fixture.componentInstance;
+
+      const clickSpy = spyOn(myComp, 'onClick');
+      const calls = profilerSpy.calls.all();
+      const button = fixture.nativeElement.querySelector('button');
+
+      button?.click();
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:104]: nit:
```suggestion
      const button = fixture.nativeElement.querySelector('button')!;

      button.click();
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+
+      expect(clickSpy).toHaveBeenCalled();
+
+      const outputStart = calls.map((call: any) => call.args)
+                              .find((arg: any) => arg[0] === ProfilerEvent.OutputStart);
+      const outputEnd = calls.map((call: any) => call.args)
+                            .find((arg: any) => arg[0] === ProfilerEvent.OutputEnd);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:111]: I'd propose creating small util function:

```
function findProfilerCall(eventType: ProfilerEvent): any[] {
  return profilerSpy.calls.all().map((call: any) => call.args).find((arg: any) => arg[0] === eventType);
}
```
and use it in all test cases like:
```suggestion
      const outputStart = findProfilerCall(ProfilerEvent.OutputStart);
      const outputEnd = findProfilerCall(ProfilerEvent.OutputEnd);
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+      const handler = new ErrorHandler();
+      const errorSpy = spyOn(handler, 'handleError');
+
+      TestBed.configureTestingModule(
+          {declarations: [MyComponent], providers: [{provide: ErrorHandler, useValue: handler}]});
+      const fixture = TestBed.createComponent(MyComponent);
+      const myComp = fixture.componentInstance;
+      const button = fixture.nativeElement.querySelector('button');
+
+      button?.click();
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:137]:
```suggestion
      const button = fixture.nativeElement.querySelector('button')!;

      button.click();
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+      const outputStart = calls.map((call: any) => call.args)
+                              .find((arg: any) => arg[0] === ProfilerEvent.OutputStart);
+      const outputEnd = calls.map((call: any) => call.args)
+                            .find((arg: any) => arg[0] === ProfilerEvent.OutputEnd);
+
+      expect(outputStart).toBeTruthy();
+      expect(outputEnd).toBeTruthy();
+
+      expect(outputStart[1]).toEqual(myComp!);
+      expect(outputEnd[1]).toEqual(myComp!);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:117]: I think we can omit first 2 asserts and just keep bottom ones:
```suggestion
      expect(outputStart?.[1]).toEqual(myComp!);
      expect(outputEnd?.[1]).toEqual(myComp!);
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+    });
+  });
+
+  describe('lifecycle hooks', () => {
+    it('should call the profiler on lifecycle execution', () => {
+      @Component({selector: 'my-comp', template: ''})
+      class MyComponent implements OnInit {
+        ngOnInit() {}
+      }
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:192]: I think it'd be great to add tests for other lifecycle hooks too. May be just expand these tests to add more hooks inside the `MyComponent` instance?

File: packages/core/src/render3/instructions/shared.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (mgechev) [reply packages/core/src/render3/instructions/shared.ts:516]: Addressed as we discussed offline. The event name is now better aligned with the purpose of the hook.

File: packages/core/src/render3/instructions/listener.ts [hunk 20]
```diff
@@ -227,13 +229,18 @@ function listenerInternal(
 }
 
 function executeListenerWithErrorHandling(
-    lView: LView, listenerFn: (e?: any) => any, e: any): boolean {
+    lView: LView, context: {}|RootContext|null, listenerFn: (e?: any) => any, e: any): boolean {
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/src/render3/instructions/listener.ts]: Yes, that is correct.

File: packages/core/src/render3/profiler.ts (hunk 14 above)
Comment (mhevery) [reply packages/core/src/render3/profiler.ts:97]: SGTM

**Review by mhevery — APPROVED** [review]

File: packages/core/src/render3/profiler.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+  if (profilerCallback != null) {
```
Comment (AndrewKushnir) [inline packages/core/src/render3/profiler.ts:98]:
```suggestion
  if (profilerCallback !== null) {
```

File: packages/core/test/acceptance/profiler_spec.ts [hunk 22]
```diff
@@ -0,0 +1,286 @@
...
+      expect(outputEnd[1]).toEqual(myComp!);
+    });
+  });
+
+  describe('lifecycle hooks', () => {
+    it('should call the profiler on lifecycle execution', () => {
+      @Component({selector: 'my-comp', template: ''})
+      class MyComponent implements OnInit, AfterViewInit, AfterViewChecked, AfterContentInit,
+                                   AfterContentChecked, DoCheck {
+        ngOnInit() {}
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:188]: We could also add `ngOnChanges` here as well.

File: packages/core/test/acceptance/profiler_spec.ts [hunk 23]
```diff
@@ -0,0 +1,286 @@
...
+      const fixture = TestBed.createComponent(MyComponent);
+
+      expect(profilerSpy).toHaveBeenCalled();
+
+      expect(profilerSpy.calls.argsFor(0)).toEqual([
+        ProfilerEvent.TemplateCreateStart, fixture.componentInstance, undefined
+      ]);
+      expect(profilerSpy.calls.argsFor(1)).toEqual([
+        ProfilerEvent.TemplateCreateEnd, fixture.componentInstance, undefined
+      ]);
```
Comment (AndrewKushnir) [inline packages/core/test/acceptance/profiler_spec.ts:59]: It looks like these calls (and similar ones below) can be refactored to use `findProfilerCall` helper function.

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: size-tracking

File: goldens/size-tracking/integration-payloads.json [hunk 24]
```diff
@@ -39,7 +39,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 2285,
-        "main-es2015": 240874,
+        "main-es2015": 240883,
```
Comment (AndrewKushnir) [inline goldens/size-tracking/integration-payloads.json:42]: We can technically revert this change (to eliminate the need to have `size-tracking` group approvals), since the delta is below the threshold (500b or 1% increase) :)

**Review by AndrewKushnir — APPROVED** [review]
Thanks for addressing the feedback @mgechev 👍 

I've added a few last comments, but they are not critical, so I LGTM the PR as well.

File: goldens/size-tracking/integration-payloads.json (hunk 24 above)
Comment (AndrewKushnir) [reply goldens/size-tracking/integration-payloads.json:42]: Actually, plz ignore this comment :) We have the necessary coverage from `size-tracking` group.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: size-tracking

File: packages/core/src/render3/profiler.ts (hunk 21 above)
Comment (mgechev) [reply packages/core/src/render3/profiler.ts:98]: I intentionally used `!=` to catch `undefined` values as well.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:365936947:BASE:365936967:1617152249708:9d4c7fbc).

**Comment by AndrewKushnir** [comment]
@mgechev I've started a presubmit and will let you know how it goes.

Just noticed that the commit message doesn't contain information that profiler is a private (and experimental) API for now. Could you please add a note that the profiler API should not be relied upon and might be changed or removed (at least from the `ng` namespace), similar to [the note you've added here](https://github.com/angular/angular/pull/41255/files#diff-eebc9cbb471b04048cba00a3579d15092638491f588a0e43473ed28594ed6345R77)?

**Comment by AndrewKushnir** [comment]
@mgechev FYI, presubmits went well for the changes in this PR. Feel free to add the "merge" label if there are no additional changes needed. Thank you.
